<launch>
    <arg name="bag_file_use" default="false" />
    <arg name="bag_file" default="/Monster/dataset/UDI_dataset/shandong_20200528/udist_four_lidar.bag" /> 

    <param name="use_sim_time" value="true" />
    <group if="$(arg bag_file_use)">
        <node pkg="rosbag" type="play" name="rosbag" args="--clock --pause -r 1 $(arg bag_file)" />
    </group>    

    <arg name="run_mloam" default="true" />
    <arg name="run_aloam" default="false" />
    <arg name="run_legoloam" default="false" />

    <arg name="data_path" default="/Monster/dataset/UDI_dataset/xfl_20200301/whole_travel/" />
    <arg name="data_source" default="bag" />
    <arg name="start_idx" default="2100"/>
    <arg name="end_idx" default="25229"/> <!-- 25229 -->
    <arg name="delta_idx" default="2" />
    <arg name="time_now" default="true" />

    <arg name="result_save" default="false"/>
    <arg name="output_path" default="/home/jjiao/trajectory_results/real_vehicle/shandong/rv01_gf/"/>
    <arg name="with_ua" default="true" /> <!-- uncertainty propagation -->    
    <arg name="gf_ratio" default="0.5" />    

    <arg name="config_file" default="$(find mloam)/config/config_realvehicle_shandong.yaml" />
    <group if="$(arg run_mloam)">
        <node pkg="mloam" type="mloam_node_rv" name="mloam_node_rv" 
            args="-config_file=$(arg config_file)
                  -data_source=$(arg data_source)
                  -result_save=$(arg result_save)
                  -output_path=$(arg output_path)
                  -data_path=$(arg data_path)
                  -start_idx=$(arg start_idx)
                  -end_idx=$(arg end_idx)
                  -delta_idx=$(arg delta_idx)
                  -time_now=$(arg time_now)">
            <remap from="/base_pose_gt" to="/current_odom"/>
        </node>
        <node pkg="mloam" type="lidar_mapper" name="lidar_mapper"
            args="-config_file=$(arg config_file)
                  -result_save=$(arg result_save) 
                  -output_path=$(arg output_path)
                  -with_ua=$(arg with_ua)
                  -gf_ratio=$(arg gf_ratio)" output="screen">
            <remap from="/laser_odom" to="/laser_odom_0"/>
        </node>
    </group>

    <!-- publish static tf -->
    <!-- x y z yaw pitch roll -->
    <!-- x y z qx qy qz qw -->
    <include file="$(find rviz_car_model)/launch/default.launch" />
    <node name="static_tf_base_sensor_into_base_link" pkg="tf" type="static_transform_publisher"
        args="0 0 -1 0 0 0 1 aft_mapped base_link 10"/>                
    <node name="static_tf_world_into_world_base_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0 -1 0 0 0 1 world world_base 10"/>
    <node name="static_tf_world_sensor_into_base_sensor" pkg="tf" type="static_transform_publisher"
        args="0 0 0 0 0 0 1 world laser_0 10"/>

    <!-- merge cloud -->
    <group unless="$(arg run_mloam)">
        <node pkg="mloam" type="test_merge_pointcloud_rv" name="test_merge_pointcloud_rv" 
            args="1 $(arg data_path) $(arg mloam_start_idx) $(arg mloam_end_idx) $(arg mloam_delta_idx)"/>      
        <node name="static_tf_world_sensor_into_camera" pkg="tf" type="static_transform_publisher"
            args="0 0 0 0 0 0 1 world camera_init 10"/>            
    </group>

    <!-- run aloam  -->
    <group if="$(arg run_aloam)">
        <include file="$(find aloam_velodyne)/launch/aloam_exp.launch">
            <arg name="aloam_odom_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_aloam_odom_estimate.txt"/>
            <arg name="aloam_map_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_aloam_map_estimate.txt"/>
        </include>       
`  </group>        

    <!-- run legoloam  -->
    <group if="$(arg run_legoloam)">
        <include file="$(find lego_loam)/launch/legoloam_exp.launch">
            <arg name="lego_loam_odom_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_legoloam_odom_estimate.txt"/>
            <arg name="lego_loam_map_path" default="/home/jjiao/trajectory_results/real_vehicle/rv01_20200301_whole_travel_test_2100_25229/stamped_legoloam_map_estimate.txt"/>
        </include>
    </group>  

 </launch>
